src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWebTemplateTest.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using Microsoft.AspNetCore.BrowserTesting;
+using Microsoft.AspNetCore.InternalTesting;
+using Templates.Test.Helpers;
+
+namespace BlazorTemplates.Tests;
+
+public class BlazorWebTemplateTest(ProjectFactoryFixture projectFactory) : BlazorTemplateTest(projectFactory)
+{
+    public override string ProjectType => "blazor";
+
+    [ConditionalTheory]
+    [SkipNonHelix]
```

Comment:
For whatever reason, the variations of this test that use WebAssembly fail in the "Windows Server x64" pipeline because the server project fails to restore:

```text
[0.001s] [TestLifetime] [Information] Starting test BlazorWebTemplate_Works-Chromium-WebAssembly at 2024-06-24T18:49:59
[1.672s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] ==> D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\dotnet\dotnet.exe new blazor --debug:disable-sdk-templates --debug:custom-hive "D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\Hives\32725d14-4041-44f9-b3f3-9874f02ba124\.templateengine" --no-restore -int WebAssembly -o D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\]
[2.548s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] The template "Blazor Web App" was created successfully.
[2.548s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] This template contains technologies from parties other than Microsoft, see https://aka.ms/aspnetcore/9.0-third-party-notices for details.
[2.806s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] 
[2.806s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] An update for template package 'Microsoft.DotNet.Web.ProjectTemplates.9.0::9.0.0-ci' is available.
[2.806s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] To update the package use:
[2.807s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information]    dotnet new install Microsoft.DotNet.Web.ProjectTemplates.9.0::9.0.0-preview.7.24321.1
[2.807s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] 
[2.826s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] Process exited.
[2.826s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] ==> D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\dotnet\dotnet.exe restore /bl [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av]
[3.719s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information]   Determining projects to restore...
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.Components.Endpoints. No packages exist with this id in source(s): D:\a\_work\1\s\artifacts\packages\Release\NonShipping\, D:\a\_work\1\s\artifacts\packages\Release\Shipping\, dotnet-eng, dotnet-public, dotnet-tools, dotnet31-transport, dotnet5-transport, dotnet6-transport, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport, dotnet9, dotnet9-transport, richnav [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102: Unable to find package Microsoft.AspNetCore.StaticFiles with version (>= 9.0.0-ci) [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 28 version(s) in dotnet-public [ Nearest version: 2.2.0 ] [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in D:\a\_work\1\s\artifacts\packages\Release\NonShipping\ [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in D:\a\_work\1\s\artifacts\packages\Release\Shipping\ [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in richnav [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.943s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet-eng [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet-tools [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet31-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet5-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet6-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet7-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet7 [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet8-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet8 [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet9-transport [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1102:   - Found 0 version(s) in dotnet9 [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[6.944s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.StaticAssets. No packages exist with this id in source(s): D:\a\_work\1\s\artifacts\packages\Release\NonShipping\, D:\a\_work\1\s\artifacts\packages\Release\Shipping\, dotnet-eng, dotnet-public, dotnet-tools, dotnet31-transport, dotnet5-transport, dotnet6-transport, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport, dotnet9, dotnet9-transport, richnav [D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.sln]
[7.089s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information]   Failed to restore D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.csproj (in 3.06 sec).
[7.592s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information]   Restored D:\a\_work\1\s\src\ProjectTemplates\test\Templates.Blazor.Tests\bin\Release\net9.0\TestTemplates\AspNet.tgiswkrhl5av\AspNet.tgiswkrhl5av.Client\AspNet.tgiswkrhl5av.Client.csproj (in 3.58 sec).
[7.660s] [BlazorTemplates.Tests.BlazorWebTemplateTest] [Information] Process exited.

```

This seemed like a build order issue, but it looked like the ref assemblies were being produced before the test project was built... so I'm not sure.

I've disabled the test in non-Helix scenarios for now (the Helix pipelines still test this on Windows), but I can open a follow-up issue to track figuring out how to enable it again.

Reply:
I have a theory about why this isn't working.

We currently have a `RemoveSharedFrameworkOnlyRefsFromNuspec` target that re-writes the `.nuspec` in each of our packages to use `<frameworkReference>` instead of `<dependency>` for references to shared framework packages.

As part of that target, the `Microsoft.AspNetCore.Components.WebAssembly.Server` package gets updated to reference `Microsoft.AspNetCore.Components.Endpoints` and `Microsoft.AspNetCore.StaticFiles` from the shared framework.

But the tests build and run _before_ that transformation happens, so nuget sources are used when attempting to restore the `M.A.C.Endpoints` and `M.A.StaticFiles` dependencies, which fails.

One way to make this work might be to build and run the template tests _after_ that target runs. It's a bit weird, since that target runs as part of `AfterSolutionBuild.proj`, which is currently the last thing to build.

Or we can ignore this problem and keep these tests Helix-only. If we do that, then we could update this PR to restrict some of the build process changes to local development scenarios only.

Reply:
@MackinnonBuck I'm more inclined to fix this than leave it there. Seems like a specific problem in the build order that might result in tests being run against the wrong assemblies in other platforms and might have other unexpected consequences in the future.

We should "act in principle" and establish that all the assets we need for the tests need to be on their "final expected form" before we try to run any test.

Reply:
To get the tests running again sooner, let's address this in a follow-up PR. I'll file an issue to track this.